cli/src/histogram.rs

```diff
@@ -0,0 +1,83 @@
+use std::{cmp, fmt};
+
+#[derive(Debug)]
+pub struct Histogram {
+    min: f64,
+    max: f64,
+    bins: usize,
+    values: Vec<usize>,
+}
+
+impl Histogram {
+    pub fn new(data: &[f64], min: f64, max: f64, bins: usize) -> Histogram {
+        let mut values: Vec<usize> = vec![0; bins];
+
+        let step = (max - min) / bins as f64;
+        let scale = 100.0;
+
+        for &y in data.iter() {
+            if y < min || y > max {
+                continue;
+            }
+
+            let mut bucket_id = ((y * scale).floor() / (step * scale)) as usize;
+
+            // Account for packages with a "perfect" (i.e. 1.0) score
+            // This is generally unlikely but possible with packages that have
+            //  not yet had analytics run on them
+            // Also account for scores on the edge 10, 20, 30...
+            if y != 0.0 && (y * 100.0) % 10.0 == 0.0 {
+                bucket_id -= 1;
+            }
+
+            if bucket_id < values.len() {
+                values[bucket_id as usize] += 1;
+            }
+        }
+        Histogram { min, max, bins, values }
+    }
+
+    fn buckets(&self) -> Vec<(f64, f64)> {
+        let step = (self.max - self.min) / self.bins as f64;
+        let mut buckets: Vec<(f64, f64)> = Vec::new();
+
+        let mut acc = self.min;
+        while acc < self.max {
+            buckets.push((acc, acc + step));
+            acc += step;
+        }
+        buckets.pop();
+        buckets
+    }
+}
+
+impl fmt::Display for Histogram {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        // Scale package count to cell width for the histogram's bar.
+        let max = *self.values.iter().max().unwrap_or(&1) as f32;
+        let scale = |count| {
+            // Show no bar for zero packages.
+            if count == 0 {
+                return 0;
+            }
+
+            let scaled = 56.0 * f32::log2(count as f32) / f32::log2(max);
+
+            // Ensure non-zero package counts ALWAYS show at least one bar.
+            cmp::max(scaled as usize, 1)
+        };
+
+        let mut histogram = String::new();
+        for (count, (min, max)) in self.values.iter().rev().zip(self.buckets().iter().rev()) {
+            let bar_min = if *min < f64::EPSILON { 0 } else { (100. * min).round() as u32 + 1 };
+            let bar_max = (100. * max).round() as u32;
+
+            let bar = "█".repeat(scale(*count));
+
+            histogram += &format!("\n{:>4} - {:<4} [{:>5}] {}", bar_min, bar_max, count, bar);
+        }
+
+        write!(f, "{:^10} {:>8}", "Score", "Count")?;
+        write!(f, "{}", histogram)
+    }
+}
```

Comment:
I basically went for a complete rewrite of this function, since (imo) the previous version was not maintenance friendly.

I hope the expanded verbosity with more descriptive intermediate variables and comments makes it much easier to understand even for people unfamiliar with Rust what is going on here.

I'd be curious what you think @maxrake.

Reply:
This re-write is much easier to follow. The descriptive names and better spacing/layout of the code sections does help readability. Ensuring that non-zero package counts always show at least one bar takes care of the bug where _no_ bar was displayed for a bucket package count of 1.

There are still some edge cases in the scaling output that might look odd, but likely not worth worrying about in this review. For instance...an extreme case is when each bucket with a package has _exactly_ 1 package:

---

<img width="965" alt="image" src="https://user-images.githubusercontent.com/18729796/181069793-20da80c6-c5a2-44d3-9093-3e2c00800e66.png">

---

The bars are all the same length...which is expected, but they are 1 bar wide instead of 56 bars wide. Similarly, the "scaling" of a bucket with a single package in it will _always_ be 1 bar wide, which is most often going to be incorrect when considering a logarithmic scale:

---

<img width="965" alt="image" src="https://user-images.githubusercontent.com/18729796/181072026-2b16c00e-db4d-4207-9ad2-044438781f9c.png">

<img width="965" alt="image" src="https://user-images.githubusercontent.com/18729796/181072482-d50bca91-c2c2-4f74-b8e1-eab21462d18b.png">

---

Again, this shouldn't be taken as reason to make further changes...just that there are some edge cases that might still be unaccounted for.

Reply:
Thanks for pointing this out. If someone has a suggestion for changing the scaling algorithm completely I'm happy to do so, but I think with this PR it's best to just focus on fixing the critical issue and keeping the underlying algorithm the same.

Reply:
Nothing comes to mind for me...and I also like keeping this PR focused on the critical issue.

That said, @kylewillmon has a history of enjoying these kinds of math/logic/algorithm problems...and usually finding elegant solutions!

Reply:
```
let scaled = 56.0 * f32::log2(count as f32) / f32::log2(max);
```

This line presents a big problem when `max` is 1 because `log2(1)` is 0.... So that becomes `NaN` and then it becomes 0 after a cast to `usize` (which I had to figure out [on the playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f83e604449ce310654f6e737cd5b7771)).

The simple solution to this is to just double the numbers before the call to `log2()`

```
let scaled = 56.0 * f32::log2(count as f32 * 2.0) / f32::log2(max * 2.0);
```

Reply:
Simple...elegant...love it!

The divide by zero issue was also observed in code earlier, but running it did not result in an error (see screenshot above) so the thought was there was some sort of short circuit in play with the numerator being 0. That playground link is cool and really helps in understanding what is really going on.

Reply:
The change works:

<img width="965" alt="image" src="https://user-images.githubusercontent.com/18729796/181112018-e8773ef6-5245-4320-890a-37a46c657b44.png">